pkg/reconciler/topology/partitionset_reconcile.go

```diff
@@ -0,0 +1,209 @@
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+
+	corev1alpha1 "github.com/kcp-dev/kcp/pkg/apis/core/v1alpha1"
+	conditionsv1alpha1 "github.com/kcp-dev/kcp/pkg/apis/third_party/conditions/apis/conditions/v1alpha1"
+	"github.com/kcp-dev/kcp/pkg/apis/third_party/conditions/util/conditions"
+	topologyv1alpha1 "github.com/kcp-dev/kcp/pkg/apis/topology/v1alpha1"
+)
+
+func (c *controller) reconcile(ctx context.Context, partitionSet *topologyv1alpha1.PartitionSet) error {
+	var labelSelector labels.Selector
+	if partitionSet.Spec.ShardSelector == nil {
+		labelSelector = labels.Everything()
+	} else {
+		var err error
+		labelSelector, err = metav1.LabelSelectorAsSelector(partitionSet.Spec.ShardSelector)
+		// This should not happen due to OpenAPI validation
+		if err != nil {
+			conditions.MarkFalse(
+				partitionSet,
+				topologyv1alpha1.PartitionSetValid,
+				topologyv1alpha1.PartitionSetInvalidSelectorReason,
+				conditionsv1alpha1.ConditionSeverityError,
+				err.Error(),
+			)
+			conditions.MarkFalse(
+				partitionSet,
+				topologyv1alpha1.PartitionsReady,
+				topologyv1alpha1.ErrorGeneratingPartitionsReason,
+				conditionsv1alpha1.ConditionSeverityError,
+				"",
+			)
+			// No need to requeue if the selector is not valid
+			return nil
+		}
+	}
+	conditions.MarkTrue(partitionSet, topologyv1alpha1.PartitionSetValid)
+
+	shards, err := c.listShards(labelSelector)
+	if err != nil {
+		conditions.MarkFalse(
+			partitionSet,
+			topologyv1alpha1.PartitionsReady,
+			topologyv1alpha1.ErrorGeneratingPartitionsReason,
+			conditionsv1alpha1.ConditionSeverityError,
+			"error listing Shards",
+		)
+		return err
+	}
+
+	oldPartitions, err := c.getPartitionsByPartitionSet(partitionSet)
+	if err != nil {
+		conditions.MarkFalse(
+			partitionSet,
+			topologyv1alpha1.PartitionsReady,
+			topologyv1alpha1.ErrorGeneratingPartitionsReason,
+			conditionsv1alpha1.ConditionSeverityError,
+			err.Error(),
+		)
+		return err
+	}
+
+	var matchLabelsMap map[string]map[string]string
+	if partitionSet.Spec.ShardSelector != nil {
+		matchLabelsMap = partition(shards, partitionSet.Spec.Dimensions, partitionSet.Spec.ShardSelector.MatchLabels)
+	} else {
+		matchLabelsMap = partition(shards, partitionSet.Spec.Dimensions, nil)
+	}
+	partitionSet.Status.Count = uint16(len(matchLabelsMap))
+	existingMatches := map[string]struct{}{}
+	var placeHolder struct{}
+	newMatchExpressions := []metav1.LabelSelectorRequirement{}
+	if partitionSet.Spec.ShardSelector != nil {
+		newMatchExpressions = partitionSet.Spec.ShardSelector.MatchExpressions
+	}
+	for _, oldPartition := range oldPartitions {
+		// MatchExpressions need to be the same
+		oldMatchExpressions := []metav1.LabelSelectorRequirement{}
+		if oldPartition.Spec.Selector != nil {
+			oldMatchExpressions = oldPartition.Spec.Selector.MatchExpressions
+		}
+		if !equality.Semantic.DeepEqual(oldMatchExpressions, newMatchExpressions) {
```

Comment:
would it make sense to change this code so that there is only one path that deletes a partition ? i.e.
```
if  !matchExpressions || !partitionKey {
 delete...
}
```

Reply:
`deletePartition` and `conditions.MarkFalse`are two lines. If I understand your point correctly, in term of readability I prefer to have the action next to the evaluation rather than differing it.